rfcs/0000-api-cr-crowding.md

```diff
@@ -0,0 +1,337 @@
+The main reason to avoid this is the data changes too frequently, on the order of once/hour, while
+we don't usually publish a feed more than once or twice a day. Plus the frequent changes would
+make it harder for clients (who aren't using `occupancies.txt` yet) to know if the data they care
+about has changed.
+
+## Small, temporary endpoint
+
+We could have done a hacky temporary `/temp_commuter_rail_crowding` endpoint. This would be low
+effort, and is appropriately quarantined from the rest of the code.
+
+However, this would set an unhappy precedent about "temporary" endpoints just for dotcom that
+don't need to coexist with the rest of the app. This flies in the face of the API's philosophy to
+this point, of being an open source app that could, in principle, be used by any transit agency,
+and that adheres to transit standards for the most part.
+
+And the most likely outcome of this "experiment" is that it lingers indefinitely. We don't want a
+"temporary", ad-hoc endpoint perpetually in the API, especially when its functionality overlaps
+with a GTFS standard that will (eventually) be accepted, and we'll want to adopt formally, for
+Transit app, Google Maps, etc.
+
+## Not exposed directly, but joined to other models
+
+Another approach is to not have an `/occupancies` endpoint but still have the model and data,
+either simply stuck in a `Trip`, or `include=occupancies` by a trip.
+
+The API has a precedent for something like this, where it digests `calendar.txt` but doesn't
+expose a `Service` resource or endpoint.
+
+Adding the given `occupancy_status` and `occupancy_percentage` values right into the trip, would
+be a straightforward approach. However, this is contrary to the data model in the current
+proposal, which explicitly rejects adding these as columns to `trips.txt` or `stop_times.txt`. The
+reasoning given is that a given trip could have different occupancy predictions for different date
+spans or days of the week. While we could handle this in the API via giving trips a _list_ of
+occupancies, this isn't feasible in `trips.txt` in GTFS. And so in the name of forward
+compatibility with the proposal once it lands, we shouldn't do it that way now.
+
+The GTFS-compatible way of attaching a list of occupancies to a trip in the API would be to still
+have the `Occupancy` resource, and allow `trips` to `include=occupancies`. I'd be open to this
+idea, if dotcom thought it would be the most natural way to work with the data, though I'm
+hesitant about it because I'm not exactly sure of the joining logic here, given the data model
+issues mentioned before. Should we validate date range and days of the week as part of the join?
+My gut feeling is that an isolated `/occupancies` endpoint would be simpler for us to build and
+maintain while the proposal is being worked on. Having to work with occupancies via trips also
+seems like it would couple client implementations between their standard trips and this
+non-standard feature more.
```

Comment:
Put this way, `/occupancies` definitely seems like a "transitional" implementation that's oriented more towards our own convenience than anything else. Occupancy data pretty much has to be connected to trips or stop times for it to be meaningful, and with this approach, clients are forced to put the pieces together themselves.

We say this is a trade-off for making the data "experimental" and subject to wholesale change or removal, but with my other comment about compatibility for this endpoint, that argument seems less clear — there's some level of support we'd have to provide in either case. _Does_ a standalone endpoint really preserve some level of flexibility that a relationship doesn't?

Reply:
> Occupancy data pretty much has to be connected to trips or stop times for it to be meaningful, and with this approach, clients are forced to put the pieces together themselves. ... Does a standalone endpoint really preserve some level of flexibility that a relationship doesn't?

I think you're right that it actually doesn't. I was thinking from the API side we would want to punt on the question of "well, should we make trip <-> occupancy 1-to-1 or 1-to-many?". And while forcing clients to go to `/occupancies` does do that, it just means the client now has to make that decision when, e.g., grouping the list by trip ID. And that's a much more subtle data dependency. Changing the data model by, e.g., adding multiple occupancies per trip via stop sequence would be just as breaking for dotcom as changing `include=occupancy` to `include=occupancies`, for example.

Combined with @thecristen thinking the `include` approach is more elegant, let's go with including via `trips` and *not* via an `/occupancies` endpoint. That would get dotcom the ability to filter by route and direction_id for free, both of which "might be useful".